### PR TITLE
[ENH] `conditional_join` can return ragged_arrays where possible - PR no. 2

### PR DESCRIPTION
**Pull request by samukweku**
This is the second part of a series of PRs that ultimately adds support for aggregations within `conditional_join`.  where possible, `ragged_arrays` can be returned to the user, either as slices or arrays of indices, which can be used in `akimbo`, or `awkward` or `pyarrow` to aggregate the data. this should be faster than materializing the entire dataframe within pandas before aggregating.

# PR Description

Please describe the changes proposed in the pull request:

- return ragged_arrays where possible with the `get_join_indices` function
- limited to single join conditions or a range join, where the columns on the right dataframe are both monotonically increasing

**This PR relates to #1269 and #1396 .**

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

**Comment by ericmjl** [comment]
🚀 Deployed on https://deploy-preview-1397--pyjanitor.netlify.app

File: janitor/functions/_numba.py [hunk 1]
```diff
@@ -166,7 +166,7 @@ def _numba_equi_join(
     # this also lets us know if there are equi matches
     keep_rows = slice_starts < slice_ends
     if not keep_rows.any():
-        return None
+        return None, None
```
Comment (ericmjl) [inline janitor/functions/_numba.py]: @samukweku do we have documentation in-line about what the tuple elements `(None, None)` semantically mean?

**Review by ericmjl — APPROVED** [review]
Pre-approving, @samukweku! Just one question re: documentation.

**Comment by ericmjl** [comment]
I also noticed that the project test coverage has significantly gone down to <90%. I might need to dig further, but this is definitely a sign that we may need to invest more in covering edge cases throughout the repo.

File: janitor/functions/_numba.py (hunk 1 above)
Comment (samukweku) [reply janitor/functions/_numba.py]: thanks for the prompt. I'll refactor the code to a clearer form
